react/src/components/ImageInstallModal.tsx

```diff
@@ -0,0 +1,197 @@
+import { useSuspendedBackendaiClient } from '../hooks';
+import { useSetBAINotification } from '../hooks/useBAINotification';
+import { usePainKiller } from '../hooks/usePainKiller';
+import BAIModal, { BAIModalProps } from './BAIModal';
+import Flex from './Flex';
+import { EnvironmentImage } from './ImageList';
+import { App, List } from 'antd';
+import _ from 'lodash';
+import { Dispatch, Key, SetStateAction } from 'react';
+import { useTranslation } from 'react-i18next';
+
+interface ImageInstallModalInterface extends BAIModalProps {
+  open: boolean;
+  onRequestClose: () => void;
+  selectedRowKeys: Key[];
+  images: EnvironmentImage[];
+  setInstallingImages: Dispatch<SetStateAction<string[]>>;
+  setSelectedRowKeys: Dispatch<SetStateAction<Key[]>>;
+}
+export default function ImageInstallModal({
+  open,
+  onRequestClose,
+  selectedRowKeys,
+  images,
+  setInstallingImages,
+  setSelectedRowKeys,
+}: ImageInstallModalInterface) {
+  const { t } = useTranslation();
+  const baiClient = useSuspendedBackendaiClient();
+  const { message } = App.useApp();
+  const { upsertNotification } = useSetBAINotification();
+  const painKiller = usePainKiller();
+
+  if (!open) return null;
+
+  if (selectedRowKeys.length === 0) {
+    message.warning(t('environment.NoImagesAreSelected'));
+    onRequestClose();
+    return null;
+  }
+
+  const mapImages = () => {
+    const imagesMappedToid = _.keyBy(images, 'id');
+    let hasInstalledImage = false;
+
+    const imagesToInstall = selectedRowKeys
+      .map((id) => imagesMappedToid[_.toString(id)])
+      .filter((image) => {
+        if (image?.installed && !hasInstalledImage) hasInstalledImage = true;
+        return !image?.installed;
+      });
+
+    return { hasInstalledImage, imagesToInstall };
+  };
+  const { hasInstalledImage, imagesToInstall } = mapImages();
+
+  if (imagesToInstall.length === 0) {
+    message.warning(t('environment.AlreadyInstalledImage'));
+    onRequestClose();
+    return null;
+  }
```

Comment:
If the modal's openness were managed inside the modal, it would consume unnecessary resources, and since it is not the role of the component, it seems better to manage it in the parent component.

Reply:
However, if the parent component manages the determination of whether an image is already installed, then mapImages() must be implemented in the parent component as well. 
What do you think about this resource?